NEW: SchemaConfig layer, configuration overhaul

unclecheese opened the pull request:
Moderate but low-impact API changes in here.

* Move almost all config to the Schema itself, not global. This was causing all kinds of issues, like [resolver pollution](https://github.com/silverstripe/silverstripe-graphql/issues/328)
* New `SchemaContext` class composed into schemas. More testable.
* `operations` now a monolith under `modelConfig`
* New `config` node for schema (`modelConfig` is nested)
* Fix bug where input types used `DBFile` (Illegal, not an input type)
* Remove `ResolverRegistry`. Just add class names to `config.resolvers`
* Add `resolverStrategy` as schema config in lieu of inheritance pattern (DefaultResolverProvider)
* Remove `DefaultResolverProvider`
* One resolver strategy per schema

## Dependent PRs ##

- [ ] CMS: https://github.com/silverstripe/silverstripe-cms/pull/2617
- [ ] Elemental: https://github.com/dnadesign/silverstripe-elemental/pull/854
- [ ] Versioned: https://github.com/silverstripe/silverstripe-versioned/pull/315
- [ ] AssetAdmin: https://github.com/silverstripe/silverstripe-asset-admin/pull/1164
- [ ] Framework (DOCS): https://github.com/silverstripe/silverstripe-framework/pull/9781

chillu reviewed (commented):
Great work!

You still have `Injector::inst()->get(FilterRegistryInterface::class)` and `Injector::inst()->get(PluginRegistry::class)` in the code (so retrieving singletons). It doesn't appear that you can register new filters or plugins *by schema*, you can only apply already registered filters/plugins to aspects of your schema. Is that correct?

chillu commented on src/Schema/Field/Field.php:
```diff
@@ -16,7 +16,6 @@
 use SilverStripe\GraphQL\Schema\Interfaces\SignatureProvider;
 use SilverStripe\GraphQL\Schema\Plugin\PluginConsumer;
 use SilverStripe\GraphQL\Schema\Registry\ResolverRegistry;
```
You missed some `ResolverRegistry` references in this class

chillu reviewed (changes requested):
Not quite sure what's going on, I can't get elemental or asset admin working with this. For example, the `readFiles` query doesn't discover the custom resolver, in the `.graphql/*` schema dump it points to the `DefaultResolver`. Running this config, which works fine without the forks.

```
{
    "name": "silverstripe/installer",
    "type": "silverstripe-recipe",
    "description": "The SilverStripe Framework Installer",
    "require": {
        "php": "^7.1 || ^8",
        "silverstripe/recipe-plugin": "^1.2",
        "silverstripe/recipe-cms": "4.x-dev",
        "silverstripe-themes/simple": "~3.2.0",
        "silverstripe/login-forms": "4.x-dev",
        "dnadesign/silverstripe-elemental": "dev-pulls/4/resolving-the-mystery as 4.x-dev",
        "silverstripe/mfa": "^4.2",
        "silverstripe/totp-authenticator": "^4.0",
        "silverstripe/graphql": "dev-pulls/master/resolving-the-mystery as 4.x-dev",
        "silverstripe/cms": "dev-pulls/4/resolving-the-mystery as 4.x-dev",
        "silverstripe/versioned": "dev-pulls/1/resolving-the-mystery as 1.x-dev",
        "silverstripe/asset-admin": "dev-pulls/1/resolving-the-mystery as 1.x-dev"
    },
    "require-dev": {
        "sminnee/phpunit": "^5.7",
        "silverstripe/graphql-devtools": "dev-master"
    },
    "extra": {
        "project-files": [
            "app/_config/*",
            ".env.example"
        ],
        "public-files": [
            "assets/*",
            "favicon.ico"
        ],
        "resources-dir": "_resources",
        "project-files-installed": [
            "app/src/Page.php",
            "app/src/PageController.php"
        ]
    },
    "config": {
        "process-timeout": 600
    },
    "prefer-stable": true,
    "minimum-stability": "dev",
    "repositories": {
        "elemental": {
            "type": "vcs",
            "url": "https://github.com/unclecheese/silverstripe-elemental.git"
        },
        "asset-admin": {
            "type": "vcs",
            "url": "https://github.com/unclecheese/silverstripe-asset-admin.git"
        },
        "versioned": {
            "type": "vcs",
            "url": "https://github.com/unclecheese/silverstripe-versioned.git"
        },
        "graphql": {
            "type": "vcs",
            "url": "https://github.com/unclecheese/silverstripe-graphql.git"
        },
        "cms": {
            "type": "vcs",
            "url": "https://github.com/unclecheese/silverstripe-cms.git"
        }
    }
}
```

chillu commented:
Tried this on a fresh install, which 

```
composer create-project silverstripe/installer test 4.x-dev
cd test
composer require --no-update silverstripe/graphql:"dev-pulls/master/resolving-the-mystery as 4.x-dev" silverstripe/cms:"dev-pulls/4/resolving-the-mystery as 4.x-dev" dnadesign/silverstripe-elemental:"dev-pulls/4/resolving-the-mystery as 4.x-dev" silverstripe/versioned:"dev-pulls/1/resolving-the-mystery as 1.x-dev" silverstripe/asset-admin:"dev-pulls/1/resolving-the-mystery as 1.x-dev"
composer config repositories.graphql vcs https://github.com/unclecheese/silverstripe-graphql.git
composer config repositories.cms vcs https://github.com/unclecheese/silverstripe-cms.git
composer config repositories.versioned vcs https://github.com/unclecheese/silverstripe-versioned.git
composer config repositories.asset-admin vcs https://github.com/unclecheese/silverstripe-asset-admin.git
composer config repositories.elemental vcs https://github.com/unclecheese/silverstripe-elemental.git
composer update -W
# Required because auto building doesn't seem to work?
vendor/bin/sake dev/graphql/build
```

When going to `admin/assets`, you get a `{"data":{"readFiles":null}}` response.

chillu commented:
Filed autobuilding as a separate bug, seems unrelated to the PR: https://github.com/silverstripe/silverstripe-graphql/issues/331

unclecheese commented:
The above issue was caused by the resolvers getting overwritten in an `array_replace_recursive` call (numeric keys not getting concatenated). 

Replaced with the idiosyncratic `mergeArray()` function from silverstripe/config.

I have also deleted almost all of the tests. They are no longer compliant with the new API, and will be refactored.

Merging based on:

* tests are now green except asset-admin (dependent PR)
* having this merged will expedite the unit testing effort (currently based on this PR)
* Other PRs have been approved
* Pretty low impact as we're pre-alpha 😉
